addons/okr-daily-script.kp/knowledge.md

```diff
@@ -0,0 +1,222 @@
+   * 1  -> only foriegn installed add-ons
+   * -1 -> only self-installed
+   * 0  -> foriegn installed and self installed
+* Each client has a daily field `has_custom_theme`.
+   * 1  -> has a custom theme
+   * -1 -> has default theme
+   * 0  -> changed from default to custom on this date
+* To facilitate total population percentages, each submission date has two static fields
+  + n_custom_theme_clients (# distinct clients on that day with a custom theme)
+  + n_clients (# distinct total clients on that date)
+
+
+```python
+%%time
+
+bucket= "telemetry-parquet/addons/agg/v1"
+default_theme_id = "{972ce4c6-7e08-4474-a285-3208198ce6fd}"
+
+print yesterday,
+
+# count of distinct client submission_date, install type
+count_by_client_day = addons\
+  .select(['client_id', 'submission_date_s3',
+        'foreign_install', 'addon_id'])\
+  .distinct()\
+  .groupBy(['client_id', 'submission_date_s3','foreign_install'])\
+  .count()
+
+# count of clients that have only foriegn_installed, only self_installed and both
+user_types = count_by_client_day\
+  .select(['client_id', 'submission_date_s3', bool_to_int('foreign_install').alias('user_type')])\
+  .groupBy(['client_id', 'submission_date_s3'])\
+  .sum('user_type')\
+  .withColumnRenamed('sum(user_type)', 'user_type')
+
+count_by_client_day = count_by_client_day.join(user_types, on=['client_id', 'submission_date_s3'])
+
+
+# does a client have a custom theme?
+# aggregate distinct values on a day, since a client could have
+# changed from default to custom
+ms_has_theme = ms.select(\
+   ms.client_id, bool_to_int(ms.active_theme.addon_id != default_theme_id).alias('has_custom_theme'))\
+  .distinct()\
+  .groupBy('client_id').sum('has_custom_theme') \
+  .withColumnRenamed('sum(has_custom_theme)', 'has_custom_theme')
+               
+
+# client_id, profile_creation_date and the earliest
+# install day for an addon
+ms_install_days = ms\
+  .select(['client_id', 'profile_creation_date', 
+           fun.explode('active_addons').alias('addons')])\
+  .groupBy(['client_id', 'profile_creation_date'])\
+  .agg(fun.min("addons.install_day").alias('min_install_day'))
+    
+
+# combine data
+current = count_by_client_day\
+  .join(ms_install_days, on='client_id', how='left')\
+  .join(ms_has_theme, on='client_id', how='left')\
+  .drop('submission_date_s3')
+
+
+# add total number of distinct clients per day
+# and total number of ditscint clients with a custom theme
+n_clients = ms.select('client_id').distinct().count()
+n_custom_themes = ms_has_theme\
+  .filter(ms_has_theme.has_custom_theme >= 0)\
+  .select('client_id').distinct().count()
+current = current\
+  .withColumn('n_custom_theme_clients', fun.lit(n_custom_themes))\
+  .withColumn('n_clients', fun.lit(n_clients))
+
+# repartition data
+current = optimize_repartition(current, record_size=38)
+
+# write to s3
+current.write.format("parquet")\
+  .save('s3n://' + bucket + '/submission_date_s3={}'.format(yesterday))
```

Comment:
Also, please add `mode="overwrite"` so that the job can be re-run without manual intervention. By default the write will fail if there is existing data.